Mathlib/CategoryTheory/Limits/HasLimits.lean

```diff
@@ -0,0 +1,1355 @@
+
+@[simp]
+theorem limit.post_post {E : Type u''} [Category.{v''} E] (H : D ⥤ E) [HasLimit ((F ⋙ G) ⋙ H)] :
+    -- H G (limit F) ⟶ H (limit (F ⋙ G)) ⟶ limit ((F ⋙ G) ⋙ H) equals 
+    -- H G (limit F) ⟶ limit (F ⋙ (G ⋙ H))
+    H.map (limit.post F G) ≫ limit.post (F ⋙ G) H = limit.post F (G ⋙ H) :=
+  by ext; erw [assoc, limit.post_π, ← H.map_comp, limit.post_π, limit.post_π]; rfl
+#align category_theory.limits.limit.post_post CategoryTheory.Limits.limit.post_post
+
+end Post
+
+theorem limit.pre_post {D : Type u'} [Category.{v'} D] (E : K ⥤ J) (F : J ⥤ C) (G : C ⥤ D)
+    [HasLimit F] [HasLimit (E ⋙ F)] [HasLimit (F ⋙ G)]
+    [HasLimit ((E ⋙ F) ⋙ G)] :-- G (limit F) ⟶ G (limit (E ⋙ F)) ⟶ limit ((E ⋙ F) ⋙ G) vs 
+            -- G (limit F) ⟶ limit F ⋙ G ⟶ limit (E ⋙ (F ⋙ G)) or
+    G.map (limit.pre F E) ≫ limit.post (E ⋙ F) G = limit.post F G ≫ limit.pre (F ⋙ G) E := by
+  ext ; erw [assoc, limit.post_π, ← G.map_comp, limit.pre_π, assoc, limit.pre_π, limit.post_π]
+#align category_theory.limits.limit.pre_post CategoryTheory.Limits.limit.pre_post
+
+open CategoryTheory.Equivalence
+
+instance hasLimitEquivalenceComp (e : K ≌ J) [HasLimit F] : HasLimit (e.functor ⋙ F) :=
+  HasLimit.mk
+    { cone := Cone.whisker e.functor (Limit.cone F)
+      isLimit := IsLimit.whiskerEquivalence (limit.isLimit F) e }
+#align category_theory.limits.has_limit_equivalence_comp CategoryTheory.Limits.hasLimitEquivalenceComp
+
+-- Porting note: testing whether this still needed 
+-- attribute [local elab_without_expected_type] inv_fun_id_assoc
+
+-- not entirely sure why this is needed
+/-- If a `E ⋙ F` has a limit, and `E` is an equivalence, we can construct a limit of `F`.
+-/
+theorem hasLimitOfEquivalenceComp (e : K ≌ J) [HasLimit (e.functor ⋙ F)] : HasLimit F := by
+  haveI : HasLimit (e.inverse ⋙ e.functor ⋙ F) := Limits.hasLimitEquivalenceComp e.symm
+  apply hasLimitOfIso (e.invFunIdAssoc F)
+#align category_theory.limits.has_limit_of_equivalence_comp CategoryTheory.Limits.hasLimitOfEquivalenceComp
+
+-- `hasLimitCompEquivalence` and `hasLimitOfCompEquivalence`
+-- are proved in `CategoryTheory/Adjunction/Limits.lean`.
+section LimFunctor
+
+variable [HasLimitsOfShape J C]
+
+section
+
+/-- `limit F` is functorial in `F`, when `C` has all limits of shape `J`. -/
+@[simps obj]
+def lim : (J ⥤ C) ⥤ C where
+  obj F := limit F
+  map α := limMap α
+  map_id F := by
+    apply Limits.limit.hom_ext; intro j
+    erw [limMap_π, Category.id_comp, Category.comp_id]
+  map_comp α β := by
+    apply Limits.limit.hom_ext; intro j
+    erw [assoc, IsLimit.fac, IsLimit.fac, ← assoc, IsLimit.fac, assoc]; rfl
+#align category_theory.limits.lim CategoryTheory.Limits.lim
+
+end
+
+variable {G : J ⥤ C} (α : F ⟶ G)
+
+-- We generate this manually since `simps` gives it a weird name.
+@[simp]
+theorem limMap_eq_limMap : lim.map α = limMap α :=
+  rfl
+#align category_theory.limits.lim_map_eq_lim_map CategoryTheory.Limits.limMap_eq_limMap
```

Comment:
I think we can now just use `@[simps]` and get rid og `limMap_eq_limMap`.

It seems that we don't have an analogue of `limMap_eq_limMap` for colimits, and that `colim` is tagged with `[simps obj]` only, which is strange IMO.

Reply:
File seems fine if I do this.